File: frontend/src/productProfile.ts
import { t } from "./i18n/translate";
import type { AppLocale } from "./i18n/types";

export type StatisticalTwinProfile =
  | "full"
  | "presentation"
  | "presentation-regression"
  | "presentation-four-domains";

const fourDomainMethodIds = new Set([
  "eda.descriptive",
  "eda.graphical_summary",
  "eda.normality",
  "hypothesis.one_sample_t",
  "hypothesis.paired_t",
  "hypothesis.two_sample_t",
  "hypothesis.one_way_anova",
  "hypothesis.equivalence_tost",
  "hypothesis.two_sample_equivalence_tost",
  "hypothesis.paired_equivalence_tost",
  "hypothesis.one_sample_wilcoxon",
  "hypothesis.mann_whitney",
  "hypothesis.kruskal_wallis",
  "categorical.one_proportion",
  "categorical.two_proportion",
  "categorical.chi_square_association",
  "regression.pearson",
  "regression.xy_correlation",
  "regression.linear_model",
  "regression.partial_least_squares",
  "regression.predict",
  "regression.predict_pasted",
  "regression.linear_model_optimizer",
]);

const presentationModuleIds = {
  presentation: ["exploration", "hypothesis"],
  "presentation-regression": ["exploration", "hypothesis", "regression"],
  "presentation-four-domains": [
    "exploration",
    "hypothesis",
    "categorical",
    "regression",
  ],
} as const;

export function resolveStatisticalTwinProfile(value: unknown): StatisticalTwinProfile {
  return value === "presentation" ||
    value === "presentation-regression" ||
    value === "presentation-four-domains"
    ? value
    : "full";
}

export function analysisModuleIdsForProfile(
  profile: StatisticalTwinProfile,
): readonly string[] | null {
  return profile === "full" ? null : presentationModuleIds[profile];
}

export function isAnalysisModuleAvailableInProfile(
  moduleId: string,
  profile: StatisticalTwinProfile,
): boolean {
  const moduleIds = analysisModuleIdsForProfile(profile);
  return moduleIds === null || moduleIds.some((candidate) => candidate === moduleId);
}

export function isAnalysisDomainAvailableInProfile(
  domainId: string,
  profile: StatisticalTwinProfile,
): boolean {
  if (profile === "full") return true;
  const domainIds =
    profile === "presentation"
      ? ["basic-exploration", "mean-equivalence"]
      : profile === "presentation-regression"
        ? [
            "basic-exploration",
            "mean-equivalence",
            "correlation-regression-prediction",
          ]
        : [
            "basic-exploration",
            "mean-equivalence",
            "proportions-categorical",
            "correlation-regression-prediction",
          ];
  return domainIds.includes(domainId);
}

export function isAnalysisMethodAvailableInProfile(
  methodId: string,
  profile: StatisticalTwinProfile,
): boolean {
  return profile !== "presentation-four-domains" || fourDomainMethodIds.has(methodId);
}

export function isPresentationProductProfile(profile: StatisticalTwinProfile): boolean {
  return profile !== "full";
}

export function presentationScopeText(
  profile: StatisticalTwinProfile,
  locale: AppLocale,
): string | null {
  if (profile === "presentation") {
    return t("profile.scopeCore", {}, locale);
  }
  if (profile === "presentation-regression") {
    return t("profile.scopeRegression", {}, locale);
  }
  if (profile === "presentation-four-domains") {
    return t("profile.scopeFourDomains", {}, locale);
  }
  return null;
}

export const statisticalTwinProfile = resolveStatisticalTwinProfile(
  import.meta.env.VITE_STATISTICAL_TWIN_PROFILE,
);

export const isPresentationProfile = isPresentationProductProfile(statisticalTwinProfile);

File: backend/app/core/product_profile.py
from fastapi import status

from app.analyses.registry import get_analysis_method
from app.api.v1.schemas.analyses import AnalysisModuleId
from app.core.config import Settings
from app.core.errors import ApiError

PRESENTATION_CORE_MODULE_IDS = frozenset(
    {
        AnalysisModuleId.EXPLORATION,
        AnalysisModuleId.HYPOTHESIS,
    }
)
PRESENTATION_REGRESSION_MODULE_IDS = frozenset(
    {
        *PRESENTATION_CORE_MODULE_IDS,
        AnalysisModuleId.REGRESSION,
    }
)
PRESENTATION_FOUR_DOMAIN_MODULE_IDS = frozenset(
    {
        AnalysisModuleId.EXPLORATION,
        AnalysisModuleId.HYPOTHESIS,
        AnalysisModuleId.CATEGORICAL,
        AnalysisModuleId.REGRESSION,
    }
)
PRESENTATION_FOUR_DOMAIN_METHOD_IDS = frozenset(
    {
        "eda.descriptive",
        "eda.graphical_summary",
        "eda.normality",
        "hypothesis.one_sample_t",
        "hypothesis.paired_t",
        "hypothesis.two_sample_t",
        "hypothesis.one_way_anova",
        "hypothesis.equivalence_tost",
        "hypothesis.two_sample_equivalence_tost",
        "hypothesis.paired_equivalence_tost",
        "hypothesis.one_sample_wilcoxon",
        "hypothesis.mann_whitney",
        "hypothesis.kruskal_wallis",
        "categorical.one_proportion",
        "categorical.two_proportion",
        "categorical.chi_square_association",
        "regression.pearson",
        "regression.xy_correlation",
        "regression.linear_model",
        "regression.partial_least_squares",
        "regression.predict",
        "regression.predict_pasted",
        "regression.linear_model_optimizer",
    }
)
PRESENTATION_MODULE_IDS_BY_PROFILE = {
    "presentation": PRESENTATION_CORE_MODULE_IDS,
    "presentation-regression": PRESENTATION_REGRESSION_MODULE_IDS,
    "presentation-four-domains": PRESENTATION_FOUR_DOMAIN_MODULE_IDS,
}


def presentation_module_ids(settings: Settings) -> frozenset[AnalysisModuleId] | None:
    return PRESENTATION_MODULE_IDS_BY_PROFILE.get(settings.product_profile)


def is_presentation_profile(settings: Settings) -> bool:
    return presentation_module_ids(settings) is not None


def presentation_profile_includes_regression(settings: Settings) -> bool:
    module_ids = presentation_module_ids(settings)
    return module_ids is not None and AnalysisModuleId.REGRESSION in module_ids


def is_four_domain_presentation_profile(settings: Settings) -> bool:
    return settings.product_profile == "presentation-four-domains"


def presentation_method_ids(settings: Settings) -> frozenset[str] | None:
    if is_four_domain_presentation_profile(settings):
        return PRESENTATION_FOUR_DOMAIN_METHOD_IDS
    module_ids = presentation_module_ids(settings)
    if module_ids is None:
        return None
    candidate_method_ids = (
        "eda.descriptive",
        "eda.graphical_summary",
        "eda.normality",
        "eda.equal_variances",
        "hypothesis.one_sample_t",
        "hypothesis.paired_t",
        "hypothesis.two_sample_t",
        "hypothesis.one_way_anova",
        "hypothesis.equivalence_tost",
        "hypothesis.two_sample_equivalence_tost",
        "hypothesis.paired_equivalence_tost",
        "hypothesis.one_sample_wilcoxon",
        "hypothesis.mann_whitney",
        "hypothesis.kruskal_wallis",
        "regression.pearson",
        "regression.xy_correlation",
        "regression.linear_model",
        "regression.partial_least_squares",
        "regression.predict",
        "regression.predict_pasted",
        "regression.linear_model_optimizer",
    )
    return frozenset(
        method_id
        for method_id in candidate_method_ids
        if (method := get_analysis_method(method_id)) is not None and method.module_id in module_ids
    )


def require_available_presentation_method(settings: Settings, method_id: str) -> None:
    method_ids = presentation_method_ids(settings)
    if method_ids is None or method_id in method_ids:
        return
    raise ApiError(
        code="presentation_profile_method_unavailable",
        message="This analysis is unavailable in the selected presentation profile.",
        status_code=status.HTTP_403_FORBIDDEN,
    )
